### PR TITLE
Add simplified mechanism for rules to log warnings and errors

### PR DESCRIPTION
**Pull request by scalvert**
Implements #1766. 

TODO:

- [ ] Ensure to add attribution for @josephdsumner as some of this code was borrowed from #1771

### CODE REVIEW

File: dev/new-rule/templates/rule.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 'use strict';
 
-const Rule = require('./base');
 const AstNodeInfo = require('../helpers/ast-node-info');
+const Rule = require('./base');
```
Comment (scalvert) [inline dev/new-rule/templates/rule.js:4]: I swapped these as they result in a lint error when the file is generated.

File: lib/rules/base.js [hunk 2]
```diff
@@ -524,12 +515,25 @@ module.exports = class Base {
...
     }
+
+    if (result.node) {
+      let node = result.node;
+
+      result.line = node.loc && node.loc.start.line;
+      result.column = node.loc && node.loc.start.column;
+      result.source = this.sourceForNode(node);
+
+      delete result.node;
```
Comment (scalvert) [inline lib/rules/base.js:534]: Realized reading #1771 that we should be deleting this so the result is stable (thanks @josephdsumner!).

File: lib/rules/base.js [hunk 3]
```diff
@@ -524,12 +515,34 @@ module.exports = class Base {
...
     if (this._filePath) {
       defaults.filePath = this._filePath;
     }
+
     if (this[MODULE_NAME]) {
       defaults.moduleId = this[MODULE_NAME];
     }
+
+    // perform the node prop expansion only if those properties don't exist in result
+    if (result.node) {
```
Comment (rwjblue) [inline lib/rules/base.js:528]: Let's leverage the `Object.assign` just below to simplify this a bit.

```js
let nodeExpansion = node && {
  line: node.loc && node.loc.start.line,
  column: node.loc && node.loc.start.column,
  source: this.sourceForNode(node),
};

let reportedResult = Object.assign({}, defaults, nodeExpansion, result);
```

Comment (scalvert) [reply lib/rules/base.js:529]: Updated with a few tweaks.

File: lib/rules/base.js [hunk 4]
```diff
@@ -520,17 +511,35 @@ module.exports = class Base {
...
+        column: node.loc && node.loc.start.column,
+        source: this.sourceForNode(node),
+      };
+
+      delete result.node;
+
+      locProps = Object.assign({}, nodeProps, result);
+    }
+
+    let reportedResult = Object.assign({}, defaults, locProps, result);
```
Comment (rwjblue) [inline lib/rules/base.js:542]: I don't _think_ we need two assign's here 🤔

**Review by rwjblue — COMMENTED** [review]
I was thinking RE: this, and it might be easiest to just land the feature now (in 2.x) then after we release 3.0.0 (and are no longer maintaining a long lived `next` branch) we can update all existing rules to use the new API. 

The main reasoning here is that changing this in all the rules is likely to make future rebase/merging efforts into `next` branch difficult...

What do you think @scalvert?

File: lib/rules/base.js (hunk 4 above)
Comment (scalvert) [reply lib/rules/base.js:542]: That's true, we can just do the single assign on 542.

**Comment by scalvert** [comment]
@rwjblue agree.

File: lib/rules/base.js (hunk 4 above)
Comment (scalvert) [reply lib/rules/base.js]: Done.

**Review by rwjblue — APPROVED** [review]

**Pull request merged**